dot/network/spam_limiter.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package network
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	lrucache "github.com/ChainSafe/gossamer/lib/utils/lru-cache"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+const MaxTimeWindow = 1 * time.Minute
+const MaxCachedPeers = 50
+const MaxCachedRequestPerPeer = 10
+
+// SpamLimiter is a rate limiter implementation designed to prevent peers from spamming requests.
+// It uses a sliding window to track the recent requests from each peer.
+// If the number of requests exceeds the set limit, the peer is considered spamming,
+// and subsequent requests will be rejected.
+type SpamLimiter struct {
+	mu         sync.Mutex
+	limits     *lrucache.LRUCache[peer.ID, *lrucache.LRUCache[common.Hash, []time.Time]]
+	maxReqs    uint32
+	windowSize time.Duration
+}
+
+// NewSpamLimiter creates a new SpamLimiter with the given maximum number of requests
+func NewSpamLimiter(maxReqs uint32, windowSize time.Duration) *SpamLimiter {
+	return &SpamLimiter{
+		limits:     lrucache.NewLRUCache[peer.ID, *lrucache.LRUCache[common.Hash, []time.Time]](MaxCachedPeers),
+		maxReqs:    maxReqs,
+		windowSize: windowSize,
+	}
+}
+
+// AddRequest adds a request to the SpamLimiter
+func (rl *SpamLimiter) AddRequest(peer peer.ID, hashedRequest common.Hash) {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	// Get or create the internal cache for the peer
+	peerCache := rl.limits.Get(peer)
+	if peerCache == nil {
+		peerCache = lrucache.NewLRUCache[common.Hash, []time.Time](MaxCachedRequestPerPeer)
+		rl.limits.Put(peer, peerCache)
+	}
+
+	// Get the timestamps for the hash
+	timestamps := peerCache.Get(hashedRequest)
+	now := time.Now()
+
+	// Filter requests that are within the time window
+	var recentRequests []time.Time
+	for _, t := range timestamps {
+		if now.Sub(t) <= rl.windowSize {
+			recentRequests = append(recentRequests, t)
+		}
+	}
+
+	// Add the current request and update the cache
+	recentRequests = append(recentRequests, now)
+	peerCache.Put(hashedRequest, recentRequests)
+}
+
+// IsLimitExceeded returns true if the limit is exceeded for the given peer and hash
+func (rl *SpamLimiter) IsLimitExceeded(peer peer.ID, hashedRequest common.Hash) bool {
```

Comment:
Should this be renamed to `AddRequestAndCheckLimit`? I think could make more sense to reduce this function to just checking the limit, without any additions. In which case could either create a wrapper function that utilized both `AddRequest` and `IsLimitExceeded`, or put the burden on the caller to do this. What do you think?

Reply:
Yeah I was thinking about it when I implemented it and I'd rather to keep them into 2 different functions. Because there could be scenarios where you want to just check without adding

Reply:
Oh yes I see, I had misunderstood a part of the code. This sounds good to me